libmamba/include/mamba/core/graph_util.hpp

```diff
@@ -7,303 +7,493 @@
 #ifndef MAMBA_CORE_GRAPH_UTIL_HPP
 #define MAMBA_CORE_GRAPH_UTIL_HPP
 
-#include <map>
+#include <algorithm>
 #include <utility>
 #include <vector>
+#include <map>
 
 namespace mamba
 {
-    // Simplified implementation of a directed graph
-    // where a path exists between each node and the
-    // first one (you can think of it as a tree with
-    // potential cycles)
 
-    template <class T>
-    class graph
+    // A sorted vector behaving like a set.
+    template <typename... Args>
+    class vector_set : private std::vector<Args...>
     {
     public:
-        using node = T;
-        using node_id = size_t;
-        using node_list = std::vector<node>;
-        using edge_list = std::vector<node_id>;
-        using adjacency_list = std::vector<edge_list>;
-        using cycle_list = std::vector<node_list>;
-
-        const node_list& get_node_list() const;
-        const edge_list& get_edge_list(node_id id) const;
-
-        node_id add_node(const node& value);
-        node_id add_node(node&& value);
+        using Base = std::vector<Args...>;
+        using typename Base::value_type;
+
+        using Base::begin;
+        using Base::end;
+        using Base::rbegin;
+        using Base::rend;
+
+        using Base::cbegin;
+        using Base::cend;
+        using Base::crbegin;
+        using Base::crend;
+
+        using Base::clear;
+        using Base::empty;
+        using Base::size;
+
+        vector_set() = default;
+        vector_set(std::initializer_list<value_type> il);
+        template <typename InputIterator>
+        vector_set(InputIterator first, InputIterator last);
+
+        bool contains(value_type const&) const;
+
+        void insert(value_type&& value);
+        void insert(value_type const& value);
```

Comment:
Agreed. However it should be a `const_iterator`, because an `iterator` may break class invariants and I don't believe implementing the machinery for a non const iterator is worth the effort at this point.

Which means I should probably remove the non-const overloads of `begin` / `end`.

Reply:
Indeed